libs/logging/src/types/base.ts

```diff
@@ -1,41 +1,133 @@
+ * levels and should not be used as an actual environment. Environments have an
+ * order with 'production' being the lowest order environment and 'never' being
+ * the highest.
+ */
 export type Environment = 'never' | 'local' | 'develop' | 'staging' | 'production';
 
+/**
+ * Defines for each log level in which environment it should still be logged.
+ * Each log level will also be logged in higher order environments, but never
+ * in lower order environments. Thus, a log level set to 'production' will be
+ * logged on any other environment, but a log level set to 'never' won't be
+ * logged at all.
+ */
 export type EnvironmentLevelMap = {
   [key in LogLevel]?: Environment;
 };
 
+/**
+ * The execution context defaults to 'browser' and switches automatically to
+ * node if the 'window' global is not available. Used to control certain
+ * features that might not be available in one of the available contexts.
+ */
 export type ExecutionContext = 'node' | 'browser';
 
+/**
+ * Defines an optional coloring function for each log level.
+ */
 export type ColorLevelMap = {
   [key in LogLevel]: (message: string) => string;
 };
 
+/**
+ * Defines a color map for each execution context.
+ */
 export type ExecutionContextColorMap = {
   [key in ExecutionContext]: ColorLevelMap;
 };
 
+/**
+ * Data available in a log callback. Contains the log level, message and any
+ * options passed to the log except for tags.
+ */
 export interface CallbackData {
+  /**
+   * Optionally provided error object. Only present if it was explicitly passed
+   * to the log, even if the level was 'error' or 'fatal'.
+   */
   error?: unknown;
+  /**
+   * Level of the log message
+   */
   level: string;
+  /**
+   * The actual message that was logged
+   */
   message: string;
+  /**
+   * Optional structured data passed to the log
+   */
   meta?: unknown;
+  /**
+   * Any optional objects passed to the log
+   */
   objects?: unknown | unknown[];
 }
 
+/**
+ * Optional callback that will be run after a message is logged. Receives the
+ * log level and message and any options passed to the log except for tags.
+ */
 export type LogCallback = (data: CallbackData) => void;
 
+/**
+ * Used to pass structured data to a callback or the sentry transport. Won't
+ * get logged in the console transport.
+ */
 export type LogMetaInformation = Record<string, unknown>;
 
+/**
+ * Tags can be used to give additional structure to logs for easier filtering
+ * or to visually identify logs of different levels with commonalities.
+ */
 export type LogTags = Record<string, string>;
 
+/**
+ * Additional options for a log call.
+ */
 export interface LogOptions {
+  /**
+   * A function run after a message is logged. Receives the log level and
+   * message and any options passed to the log except for tags.
+   */
   callback?: LogCallback;
+  /**
+   * Used to pass structured data to a callback or the sentry transport. Won't
+   * get logged in the console transport.
+   */
```

Comment:
Is it possible to remove this kind of duplicate JSDocs and still be possible to retrieve "Quick Documentation" help inside VSCode/IntelliJ because the type itself is documented.

Reply:
As far as I could see, IntelliJ will not pull the JSDocs from the type to the attribute of e.g. LogOptions, which does sort of make sense since the documentation could very well be different for the type and the attribute. 
We could either try to be more specific in our documentation of type and attribute to eliminate duplication, remove the JSDoc in one place or simply live with the duplication. I guess this is open for debate.